src/niri.rs

```diff
@@ -1200,15 +1201,71 @@ impl State {
                         }
                     }
                 }
-                niri_ipc::OutputAction::Vrr { enable } => {
-                    config.variable_refresh_rate = enable;
+                niri_ipc::OutputAction::Vrr { vrr } => {
+                    config.variable_refresh_rate = match vrr {
+                        niri_ipc::VrrToSet::AlwaysOff => None,
+                        niri_ipc::VrrToSet::AlwaysOn => Some(false),
+                        niri_ipc::VrrToSet::OnDemand => Some(true),
+                    }
+                    .into()
                 }
             }
         }
 
         self.reload_output_config();
     }
 
+    pub fn refresh_on_demand_vrr(&mut self) {
+        let config = self.niri.config.borrow();
+        let output_iter: Vec<Output> = self
+            .niri
+            .output_by_name
+            .iter()
+            .filter(|(name, _)| {
+                config
+                    .outputs
+                    .find(name)
+                    .is_some_and(|output| output.variable_refresh_rate.on_demand())
+            })
+            .map(|(_, o)| o.clone())
+            .collect();
+        drop(config);
+
+        for output in &output_iter {
+            let last = self.niri.output_state[output].frame_clock.vrr();
+
+            if let Some(workspace) = self
+                .niri
+                .layout
+                .monitor_for_output(output)
+                .map(|monitor| monitor.active_workspace_ref())
+            {
+                let Some(output_size) = output.current_mode().map(|mode| mode.size) else {
+                    continue;
+                };
+                let output_geo = Rectangle::from_loc_and_size(
+                    (0, 0),
+                    output
+                        .current_transform()
+                        .invert()
+                        .transform_size(output_size),
+                );
+                let current = workspace.windows().any(|mapped| {
+                    mapped.rules().variable_refresh_rate == Some(true)
+                        && mapped
+                            .window
+                            .geometry()
+                            .to_physical_precise_round(output.current_scale().fractional_scale())
+                            .overlaps(output_geo)
```

Comment:
Window geometry is relative to the window buffer, not a global coordinate system, so this check doesn't work properly.

Either way, I feel like this should check if any surface of the window has this output as its primary output. That will deal with subsurfaces outside the window, and when we get floating, it will also deal with cases like a few pixels of a window overlapping one output (in which case we don't want to enable VRR there). It already works for things like frame callbacks, I think it would be appropriate if the VRR window rule used the same logic.

Reply:
I will have to have a deeper look into the frame callback logic. One issue is that the window may not be in the active workspace and secondly it could be not visible.